server/src/main/java/org/opensearch/search/startree/filter/provider/StarTreeFilterProvider.java

```diff
@@ -156,4 +163,176 @@ public StarTreeFilter getFilter(SearchContext context, QueryBuilder rawFilter, C
 
     }
 
+    class BoolStarTreeFilterProvider implements StarTreeFilterProvider {
+        @Override
+        public StarTreeFilter getFilter(SearchContext context, QueryBuilder rawFilter, CompositeDataCubeFieldType compositeFieldType)
+            throws IOException {
+            BoolQueryBuilder boolQueryBuilder = (BoolQueryBuilder) rawFilter;
+            if (!boolQueryBuilder.hasClauses()) {
+                return null;
+            }
+
+            if (boolQueryBuilder.must().isEmpty() == false) {
+                return processMustClauses(boolQueryBuilder.must(), context, compositeFieldType);
+            }
+
+            if (boolQueryBuilder.should().isEmpty() == false) {
+                return processShouldClauses(boolQueryBuilder.should(), context, compositeFieldType);
+            }
+
+            return null;
+        }
+
+        private StarTreeFilter processMustClauses(
+            List<QueryBuilder> mustClauses,
+            SearchContext context,
+            CompositeDataCubeFieldType compositeFieldType
+        ) throws IOException {
+            if (mustClauses.isEmpty()) {
+                return null;
+            }
+            Map<String, List<DimensionFilter>> dimensionToFilters = new HashMap<>();
+
+            for (QueryBuilder clause : mustClauses) {
+                StarTreeFilter clauseFilter;
+
+                if (clause instanceof BoolQueryBuilder) {
+                    BoolQueryBuilder boolClause = (BoolQueryBuilder) clause;
+                    if (boolClause.must().isEmpty() == false) {
+                        // Process nested MUST
+                        clauseFilter = processMustClauses(boolClause.must(), context, compositeFieldType);
+                    } else if (boolClause.should().isEmpty() == false) {
+                        // Process SHOULD inside MUST
+                        clauseFilter = processShouldClauses(boolClause.should(), context, compositeFieldType);
+                        // Note: clauseFilter now contains all SHOULD conditions as separate filters
+                    } else {
+                        return null;
+                    }
+                } else {
+                    // Only allow term, terms, and range queries
```

Comment:
nit: can be replaced with `Only allow other supported QueryBuilders`

Reply:
Done